src/lib.rs

```diff
@@ -129,6 +129,106 @@ impl<'a> From<PrivatePkcs8KeyDer<'a>> for PrivateKeyDer<'a> {
     }
 }
 
+impl<'a> TryFrom<&'a [u8]> for PrivateKeyDer<'a> {
+    type Error = &'static str;
+
+    fn try_from(key: &'a [u8]) -> Result<Self, Self::Error> {
+        const SHORT_FORM_LEN_MAX: u8 = 128;
+        const TAG_SEQUENCE: u8 = 0x30;
+        const TAG_INTEGER: u8 = 0x02;
+
+        // We expect all key formats to begin with a SEQUENCE, which requires at least 2 bytes
+        // in the short length encoding.
+        if key.first() != Some(&TAG_SEQUENCE) || key.len() < 2 {
+            return Err(INVALID_KEY_DER_ERR);
+        }
+
+        // The length of the SEQUENCE is encoded in the second byte. We must skip this many bytes.
+        let skip_len = match key[1] >= SHORT_FORM_LEN_MAX {
+            // 1 byte for SEQUENCE tag, 1 byte for short-form len
+            false => 2,
+            // 1 byte for SEQUENCE tag, 1 byte for start of len, remaining bytes encoded
+            // in key[1].
+            true => 2 + (key[1] - SHORT_FORM_LEN_MAX) as usize,
+        };
+        let key_bytes = key.get(skip_len..).ok_or(INVALID_KEY_DER_ERR)?;
+
+        // PKCS#8 (https://www.rfc-editor.org/rfc/rfc5208) describes the PrivateKeyInfo
+        // structure as:
+        //   PrivateKeyInfo ::= SEQUENCE {
+        //      version Version,
+        //      privateKeyAlgorithm AlgorithmIdentifier {{PrivateKeyAlgorithms}},
+        //      privateKey PrivateKey,
+        //      attributes [0] Attributes OPTIONAL
+        //   }
+        // PKCS#5 (https://www.rfc-editor.org/rfc/rfc8018) describes the AlgorithmIdentifier
+        // as a SEQUENCE.
+        //
+        // Therefore, we consider the outer SEQUENCE, a Version of 0, and the start of
+        // an AlgorithmIdentifier to be enough to identify a PKCS#8 key. If it were PKCS#1,
+        // the version would not be followed by a SEQUENCE. If it were SEC1, the version would
+        // not have been 0.
+        let pkcs8_hint = [TAG_INTEGER, 0x01, 0x00, TAG_SEQUENCE];
```

Comment:
All things considered, I would probably structure this code to keep the hint definitions and their usage closer together. So maybe just:

```rust
if key_bytes.starts_with([TAG_INTEGER, 0x01, 0x00, TAG_SEQUENCE]) {
     return Ok(Self::Pkcs8(key.into());
}
```

And similar for the other variants. This keeps the documentation and all of the related code closer together, which IMO will make this easier to follow.

Reply:
I implemented this feedback and agree it's much easier to follow :+1: